### ISSUE TITLE
Improve S2259: Support DoesNotReturnIf for custom assertions

### ISSUE THREAD

**Issue by llfab** [SonarSource/sonar-dotnet#5824]
### Description

S2259 is raised even if tested before with custom assertion method. This issue therefore is similar to https://github.com/SonarSource/sonar-dotnet/issues/349 with the difference that I am referring to a custom method.

### Repro steps

``` csharp
Assert.That(systems != null, "Cannot get pose for coordinate key - no coordinate systems defined!");
return systems.GetPose(coordinateKey.Value, Core.Geometry3D.CoordinateKey.World) * Pose;

//...
// with assertion method
public static void That(bool condition, string message = "Precondition failed!")
{
  if (!condition) { throw new AssertionException(message); }
}
```

### Expected behavior

The sonar C# scanner should detect that the custom assertion will check for != null. 

### Actual behavior

The sonar C# scanner does not detect and raise an issue both in SonarCloud and in SonarLint.

### Known workarounds

None. I tried the ```AssertionMethodAttribute```, but that did not work.

``` csharp
public class AssertionMethodAttribute : Attribute { }

//...
[AssertionMethod]
public static void That(bool condition, string message = "Precondition failed!")
{
    if (!condition) { throw new AssertionException(message); }
}


```

**Comment by llfab** [comment]
Please note that a simple Attribute like ```ValidatedNotNullAttribute``` would not fix the problem even if it worked because the custom assertion is able to check any kind of boolean construction.

**Comment by llfab** [comment]
I found one workaround which is very specific however:

```csharp
Assert.NotNull(systems, "Text");
return systems.GetPose(coordinateKey.Value, Core.Geometry3D.CoordinateKey.World) * Pose;

//...
public class AssertionMethodAttribute : Attribute { }
public sealed class ValidatedNotNullAttribute : Attribute { }

//...
[AssertionMethod]
public static void NotNull([ValidatedNotNull] object obj, string message = "Object must not be null!") => Assert.That(obj != null, message);
```

The (more generic) AssertionMethodAttribute did not help. The more specific ValidateNotNullAttribute did help. 

I would be better however if this was doable also with more general means (like AssertionMethodAttribute).

**Comment by pavel-mikula-sonarsource** [comment]
This is an interesting scenario. Thank you for reporting it.

Methods decorated with `AssertionMethodAttribute` should assume that their first bool argument is always true. And try to  learn constraints from the expression that is passed into it. We should be able to support the inline expressions, however, we will not be able to backtrack usages like
```
var result = systems != null;
Assert.That(result, "Something");
```

**Comment by martin-strecker-sonarsource** [comment]
The attribute @llfab is talking about is most likely the [AssertionMethodAttribute from ReSharper](https://www.jetbrains.com/help/resharper/Reference__Code_Annotation_Attributes.html#AssertionMethodAttribute). It goes in combination with [AssertionConditionAttribute﻿](https://www.jetbrains.com/help/resharper/Reference__Code_Annotation_Attributes.html#AssertionConditionAttribute). We should also support [Debug.Assert](https://docs.microsoft.com/dotnet/api/system.diagnostics.debug.assert) from the runtime, even though that one is not annotated.

**Comment by martin-strecker-sonarsource** [comment]
see also #349

**Comment by pavel-mikula-sonarsource** [comment]
@llfab This will be part of the next analyzer release 8.47. You can annotate your parameter with `DoesNotReturnIf(bool)` attribute to let the engine learn from it.

You can either reuse [System.Diagnostics.CodeAnalysisDoesNotReturnIfAttribute](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.doesnotreturnifattribute?view=net-6.0) if available, or declare your own attribute with the same signature and name.
